### PR TITLE
Adds possibility query Customer.io API for list of clients by email

### PR DESCRIPTION
**Pull request by kirill-deriv**

### CODE REVIEW

File: lib/WebService/Async/CustomerIO.pm [hunk 1]
```diff
@@ -316,29 +338,61 @@ sub add_to_segment {
...
+    return $self->tracking_request(
+        POST => "segments/$segment_id/remove_customers",
+        {ids => $customers_ids});
 }
 
+=head2 get_customers_by_email
+
+Query Customer.io API for list of clients, who has requested email address.
+
+usage: c<< get_customers_by_email($email)->future([$customer_obj1, ...]) >>
```
Comment (billmarriott-deriv) [inline lib/WebService/Async/CustomerIO.pm:373]: Please update the POD at L108-109

File: lib/WebService/Async/CustomerIO.pm [hunk 2]
```diff
@@ -316,29 +338,61 @@ sub add_to_segment {
...
 
+=cut
+
+sub get_customers_by_email {
+    my ($self, $email) = @_;
+
+    Carp::croak 'Missing required argument: email' unless $email;
+
+    return $self->api_request(GET => "customers?email=$email")->then(
```
Comment (billmarriott-deriv) [inline lib/WebService/Async/CustomerIO.pm]: Do we need to URI escape $email? i.e. `URI::Escape::uri_escape_utf8`.

File: lib/WebService/Async/CustomerIO.pm [hunk 3]
```diff
@@ -49,14 +50,16 @@ Parameters:
...
+
 =back
 
 =cut
 
 sub _init {
     my ($self, $args) = @_;
 
-    for my $k (qw(site_id api_key)) {
+    for my $k (qw(site_id api_key api_token)) {
```
Comment (mahrous-deriv) [inline lib/WebService/Async/CustomerIO.pm:62]: A Customer.IO API Token is required to consume `Regular API` endpoint with Bearer authentication.

File: lib/WebService/Async/CustomerIO.pm [hunk 4]
```diff
@@ -128,9 +139,12 @@ Usage: C<< api_request($method, $uri, $data) -> future($data) >>
...
     my ($self, $method, $uri, $data) = @_;
 
-    return $self->api_ratelimiter->acquire->then(sub {
-        $self->_request($method, join(q{/} => (API_END_POINT, $uri)), $data);
-    });
+    Carp::croak('API token is missed') unless $self->api_token;
+
+    return $self->api_ratelimiter->acquire->then(
+        sub {
+            $self->_request($method, join(q{/} => (API_END_POINT, $uri)), $data, {authorization => 'Bearer ' . $self->api_token},);
```
Comment (mahrous-deriv) [inline lib/WebService/Async/CustomerIO.pm:146]: Handle `Bearer` authorization

File: lib/WebService/Async/CustomerIO.pm [hunk 5]
```diff
@@ -189,56 +196,70 @@ my %PATTERN_FOR_ERROR = (
 );
 
 sub _request {
```
Comment (mahrous-deriv) [inline lib/WebService/Async/CustomerIO.pm:198]: Tidy only.

File: lib/WebService/Async/CustomerIO.pm [hunk 6]
```diff
@@ -189,56 +196,70 @@ my %PATTERN_FOR_ERROR = (
...
+                my $response_data = decode_json_utf8($response->content);
+                return Future->done($response_data);
+            } catch {
+                return Future->fail('UNEXPECTED_RESPONSE_FORMAT', 'customerio', $@, $response);
+            }
+        });
 }
 
-
 sub _ua {
```
Comment (mahrous-deriv) [inline lib/WebService/Async/CustomerIO.pm:253]: Tidy only

File: lib/WebService/Async/CustomerIO.pm [hunk 7]
```diff
@@ -299,9 +320,12 @@ Usage: C<< new_customer(%params) -> obj >>
...
+    my ($self, %params) = @_;
 
-    return $self->tracking_request(POST => 'events', \%params);
+    return $self->tracking_request(
+        POST => 'events',
+        \%params
+    );
 }
 
 =head2 add_to_segment
```
Comment (mahrous-deriv) [inline lib/WebService/Async/CustomerIO.pm:331]: Tidy

File: lib/WebService/Async/CustomerIO.pm [hunk 8]
```diff
@@ -316,29 +340,61 @@ sub add_to_segment {
...
 
+=head2 get_customers_by_email
+
+Query Customer.io API for list of clients, who has requested email address.
+
+usage: c<< get_customers_by_email($email)->future([$customer_obj1, ...]) >>
 
+=cut
+
+sub get_customers_by_email {
```
Comment (mahrous-deriv) [inline lib/WebService/Async/CustomerIO.pm:377]: New sub

File: lib/WebService/Async/CustomerIO.pm [hunk 9]
```diff
@@ -316,29 +340,61 @@ sub add_to_segment {
...
 
-Remove people from a manual segment.
+remove people from a manual segment.
 
-Usage: C<< remove_from_segment($segment_id, @$customer_ids) -> Future() >>
+usage: c<< remove_from_segment($segment_id, @$customer_ids) -> future() >>
 
 =cut
 
 sub remove_from_segment {
```
Comment (mahrous-deriv) [inline lib/WebService/Async/CustomerIO.pm:358]: Tidy only

File: t/api_client.t [hunk 10]
```diff
@@ -16,80 +16,111 @@ subtest 'Creating API client' => sub {
...
+        my $err = exception { WebService::Async::CustomerIO->new(%{$test_case->[0]}) };
         like $err, $test_case->[1], "Got Expected error";
     }
 
-    ok(WebService::Async::CustomerIO->new(site_id => 1, api_key => 1), 'Api Client created');
+    ok(
+        WebService::Async::CustomerIO->new(
+            site_id   => 1,
+            api_key   => 1,
+            api_token => 1
```
Comment (mahrous-deriv) [inline t/api_client.t:48]: New param `api_token`

File: t/api_client.t [hunk 11]
```diff
@@ -16,80 +16,111 @@ subtest 'Creating API client' => sub {
...
 subtest 'Getters methods' => sub {
 
     my $test_loop = IO::Async::Loop->new;
-    my $api = WebService::Async::CustomerIO->new(
-        site_id => 'some_site_id',
-        api_key => 'some_api_key',
+    my $api       = WebService::Async::CustomerIO->new(
+        site_id   => 'some_site_id',
+        api_key   => 'some_api_key',
+        api_token => 'some_api_token',
```
Comment (mahrous-deriv) [inline t/api_client.t:60]: New param `api_token`

File: t/api_client.t [hunk 12]
```diff
@@ -107,86 +138,94 @@ subtest 'API error handlig' => sub {
...
     }
 };
 
 subtest 'Handling no JSON respose' => sub {
     my $api = WebService::Async::CustomerIO->new(
-        site_id => 'some_site_id',
-        api_key => 'some_api_key',
+        site_id   => 'some_site_id',
+        api_key   => 'some_api_key',
+        api_token => 'some_api_token',
```
Comment (mahrous-deriv) [inline t/api_client.t:168]: New param `api_token`

File: t/api_client.t [hunk 13]
```diff
@@ -107,86 +138,94 @@ subtest 'API error handlig' => sub {
...
     is $err, 'UNEXPECTED_RESPONSE_FORMAT', "Got error for non JSON response";
 };
 
 subtest 'Emit anonymous event' => sub {
     my $api = WebService::Async::CustomerIO->new(
-        site_id => 'some_site_id',
-        api_key => 'some_api_key',
+        site_id   => 'some_site_id',
+        api_key   => 'some_api_key',
+        api_token => 'some_api_token',
```
Comment (mahrous-deriv) [inline t/api_client.t:192]: New param `api_token`

File: t/api_client.t [hunk 14]
```diff
@@ -107,86 +138,94 @@ subtest 'API error handlig' => sub {
...
     is_deeply $response->{data}, {some => 'data'}, 'Data is correct';
 };
 
 subtest 'Adding users to segment' => sub {
     my $api = WebService::Async::CustomerIO->new(
-        site_id => 'some_site_id',
-        api_key => 'some_api_key',
+        site_id   => 'some_site_id',
+        api_key   => 'some_api_key',
+        api_token => 'some_api_token',
```
Comment (mahrous-deriv) [inline t/api_client.t:214]: New param `api_token`

File: t/api_client.t [hunk 15]
```diff
@@ -197,28 +236,58 @@ subtest 'Adding users to segment' => sub {
...
+    is $response->{method}, 'POST',                        'Method is correct';
+    is $response->{uri},    'segments/1/remove_customers', 'URI is correct';
     is_deeply $response->{data}, {ids => [1]}, 'Data is correct';
     my $err = exception { $api->remove_from_segment(undef, [1])->get };
     like $err, qr/^Missing required attribute: segment_id/, "Got error for missing segment id";
     $err = exception { $api->remove_from_segment(1)->get };
     like $err, qr/^Invalid value for customers_ids/, "Got error for missing customer ids";
 };
 
+subtest 'Search customers by email' => sub {
```
Comment (mahrous-deriv) [inline t/api_client.t:263]: New test case for querying customers by email.

**Pull request merged**